File: lua/features.lua
-- Zoom SDK feature factory

local BaseFeature = require("feature.base_feature")
local DebugFeature = require("feature.debug_feature")
local IdempotencyFeature = require("feature.idempotency_feature")
local MetricsFeature = require("feature.metrics_feature")
local PagingFeature = require("feature.paging_feature")
local RatelimitFeature = require("feature.ratelimit_feature")
local RetryFeature = require("feature.retry_feature")
local TestFeature = require("feature.test_feature")
local TimeoutFeature = require("feature.timeout_feature")


local features = {}

features.base = function()
  return BaseFeature.new()
end

features["debug"] = function()
  return DebugFeature.new()
end

features["idempotency"] = function()
  return IdempotencyFeature.new()
end

features["metrics"] = function()
  return MetricsFeature.new()
end

features["paging"] = function()
  return PagingFeature.new()
end

features["ratelimit"] = function()
  return RatelimitFeature.new()
end

features["retry"] = function()
  return RetryFeature.new()
end

features["test"] = function()
  return TestFeature.new()
end

features["timeout"] = function()
  return TimeoutFeature.new()
end


return features

File: php/features.php
<?php
declare(strict_types=1);

// Zoom SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class ZoomFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new ZoomBaseFeature();
            case "debug":
                return new ZoomDebugFeature();
            case "idempotency":
                return new ZoomIdempotencyFeature();
            case "metrics":
                return new ZoomMetricsFeature();
            case "paging":
                return new ZoomPagingFeature();
            case "ratelimit":
                return new ZoomRatelimitFeature();
            case "retry":
                return new ZoomRetryFeature();
            case "test":
                return new ZoomTestFeature();
            case "timeout":
                return new ZoomTimeoutFeature();
            default:
                return new ZoomBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}
